File: scripts/fetch_trending.py
"""
fetch_trending.py

Busca os últimos N_DAYS do bonfy/github-trending, filtra Python/JavaScript,
mescla com o dataset existente (docs/data/repos.json) e regenera os artefatos
web (docs/data.js, docs/index.html).

Uso:  python scripts/fetch_trending.py [--days 15]
"""

import json
import re
import sys
import subprocess
import argparse
import urllib.request
import urllib.error
from datetime import date, timedelta
from pathlib import Path
from collections import Counter

ROOT       = Path(__file__).parent.parent
REPOS_JSON = ROOT / "data" / "repos.json"   # fonte de dados do projeto
DOCS_JSON  = ROOT / "docs" / "data" / "repos.json"  # mirror para GitHub Pages

ALLOWED_LANGS = {"python", "javascript"}

# Mapeamento para o formato canônico do projeto
LANG_CANONICAL = {"python": "Python", "javascript": "Javascript"}

# ── categorias (idêntico ao generate_data.py para consistência) ──────────────
CATEGORIES = [
    ("Agentic AI", [
        "agent", "crewai", "autogen", "langgraph", "mcp", "tool-calling",
        "tool_calling", "autonomous", "multi-agent", "multiagent", "agentic",
    ]),
    ("LLMs & Models", [
        "llm", "gpt", "claude", "llama", "mistral", "gemini", "fine-tun",
        "finetun", "gguf", "ollama", "ggml", "transformers", "huggingface",
        "openai", "chatgpt", "anthropic", "deepseek", "qwen", "falcon",
        "vllm", "lora", "rlhf",
    ]),
    ("Image & Video", [
        "diffusion", "comfyui", "stable-diffusion", "stablediffusion",
        "image generation", "text-to-image", "text-to-video", "animate",
        "gan", "inpaint", "controlnet", "dreambooth", "video generation", "img2img",
    ]),
    ("Audio & Voice", [
        "tts", "whisper", "speech", "voice", "audio", "music",
        "text-to-speech", "stt", "asr", "transcri", "vocal",
    ]),
    ("Data & Analytics", [
        "pandas", "etl", "dbt", "kafka", "airflow", "polars", "spark",
        "databricks", "bigquery", "redshift", "snowflake", "pyspark",
        "data analysis", "data cleaning", "sql", "warehouse", "analytics",
        "matplotlib", "plotly", "jupyter",
    ]),
    ("Security", [
        "pentest", "osint", "ctf", "vulnerability", "forensic", "exploit",
        "hacking", "malware", "reverse engineer", "cybersecurity", "infosec",
        "owasp", "xss", "injection", "zero-day",
    ]),
    ("Games & Creative", [
        "game", "3d", "pixel", "canvas", "shader", "graphics",
        "opengl", "unity", "godot", "pygame", "raylib", "threejs", "creative coding",
    ]),
    ("Self-hosted", [
        "self-host", "selfhost", "homelab", "home assistant", "nas",
        "nextcloud", "jellyfin", "plex", "immich", "download manager",
        "paperless", "torrent", "media server",
    ]),
    ("Automation & Bots", [
        "bot", "automat", "workflow", "n8n", "selenium", "playwright",
        "rpa", "scraper", "crawler", "zapier", "webhook", "scheduler",
    ]),
    ("Learning", [
        "tutorial", "course", "roadmap", "awesome", "interview", "book",
        "guide", "learn", "beginner", "30-days", "cheatsheet", "primer",
        "from scratch", "algorithms", "data structure", "system design", "leetcode",
    ]),
    ("DevTools", [
        "cli", "docker", "kubernetes", "k8s", "lint", "test", "debug",
        "monitor", "infra", "devops", "terraform", "helm", "ansible",
        "formatter", "linter", "vscode", "vim", "neovim", "shell", "zsh",
        "ci/cd", "github action", "build tool",
    ]),
    ("Web & APIs", [
        "fastapi", "django", "flask", "react", "next.js", "nextjs", "vue",
        "svelte", "angular", "graphql", "nuxt", "tailwind", "bootstrap",
        "express", "nodejs", "rest api", "frontend", "backend", "fullstack",
        "typescript", "web framework",
    ]),
]
OTHER = "Other"

def categorise(name: str, desc: str) -> str:
    text = (name + " " + desc).lower()
    for label, kws in CATEGORIES:
        if any(kw in text for kw in kws):
            return label
    return OTHER

# ── fetch bonfy raw file ──────────────────────────────────────────────────────
# 2026 files are at root; earlier years are in YYYY/ subfolders.
def bonfy_url(d: date) -> str:
    date_str = d.strftime("%Y-%m-%d")
    if d.year >= 2026:
        return f"https://raw.githubusercontent.com/bonfy/github-trending/master/{date_str}.md"
    return f"https://raw.githubusercontent.com/bonfy/github-trending/master/{d.year}/{date_str}.md"

def fetch_raw(url: str) -> str | None:
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "fetch_trending/1.0"})
        with urllib.request.urlopen(req, timeout=15) as resp:
            if resp.status == 200:
                return resp.read().decode("utf-8")
    except urllib.error.HTTPError as e:
        if e.code != 404:
            print(f"  HTTP {e.code}: {url}", file=sys.stderr)
    except Exception as e:
        print(f"  Erro ao buscar {url}: {e}", file=sys.stderr)
    return None

# ── parser do formato bonfy ───────────────────────────────────────────────────
# Cada repo: * [owner / repo](url):descrição
REPO_RE = re.compile(r"^\* \[([^\]]+)\]\(([^)]+)\)(?::(.*))?$")

def parse_bonfy_md(content: str) -> list[dict]:
    repos = []
    current_lang = None
    for line in content.splitlines():
        line = line.rstrip()
        if line.startswith("#### "):
            current_lang = line[5:].strip().lower()
        elif line.startswith("* [") and current_lang in ALLOWED_LANGS:
            m = REPO_RE.match(line)
            if not m:
                continue
            # "owner / repo" → "owner/repo"
            raw_name = re.sub(r"\s*/\s*", "/", m.group(1).strip())
            url  = m.group(2).strip()
            desc = (m.group(3) or "").strip()
            repos.append({
                "name": raw_name,
                "url":  url,
                "desc": desc,
                "lang": LANG_CANONICAL[current_lang],
            })
    return repos

# ── main ──────────────────────────────────────────────────────────────────────
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--days", type=int, default=15,
                        help="Quantos dias buscar do bonfy (default: 15)")
    args = parser.parse_args()

    today = date.today()
    dates = [today - timedelta(days=i) for i in range(args.days)]

    # 1) Buscar bonfy ----------------------------------------------------------
    print(f"Buscando {args.days} dias do bonfy/github-trending...")
    # url→{name, url, lang, desc, appearances}
    bonfy_map: dict[str, dict] = {}
    fetched = 0
    for d in reversed(dates):   # cronológico (mais antigo → mais novo)
        url = bonfy_url(d)
        content = fetch_raw(url)
        if content is None:
            print(f"  {d}  404/erro — ignorado")
            continue
        repos = parse_bonfy_md(content)
        fetched += 1
        print(f"  {d}  {len(repos):>3} repos (Python/JS)")
        for r in repos:
            key = r["url"]
            if key in bonfy_map:
                bonfy_map[key]["appearances"] += 1
                # Atualiza desc se vier preenchida
                if r["desc"] and not bonfy_map[key]["desc"]:
                    bonfy_map[key]["desc"] = r["desc"]
            else:
                bonfy_map[key] = {**r, "appearances": 1}

    print(f"\nTotal de datas buscadas: {fetched}/{args.days}")
    print(f"Repos únicos Python/JS no período: {len(bonfy_map)}")

    # Filtra explicitamente Go/Swift (salvaguarda)
    go_swift_skipped = sum(
        1 for r in bonfy_map.values()
        if r["lang"] not in ("Python", "Javascript")
    )
    if go_swift_skipped:
        print(f"Repos Go/Swift ignorados: {go_swift_skipped}")
    bonfy_map = {k: v for k, v in bonfy_map.items()
                 if v["lang"] in ("Python", "Javascript")}

    # 2) Carregar dataset existente -------------------------------------------
    existing: dict[str, dict] = {}
    if REPOS_JSON.exists():
        for r in json.loads(REPOS_JSON.read_text(encoding="utf-8")):
            existing[r["url"]] = r
    print(f"Repos existentes no dataset: {len(existing)}")

    # 3) Mesclar ---------------------------------------------------------------
    added   = 0
    updated = 0
    for url, bonfy in bonfy_map.items():
        app = bonfy["appearances"]
        if url in existing:
            # Repo já existe: incrementa dias e atualiza desc se vazia
            existing[url]["days"] = existing[url].get("days", 0) + app
            if bonfy["desc"] and not existing[url].get("desc"):
                existing[url]["desc"] = bonfy["desc"]
            updated += 1
        else:
            # Repo novo
            rating = "High" if app >= 5 else ("Medium" if app >= 2 else "Low")
            existing[url] = {
                "name":  bonfy["name"],
                "url":   url,
                "lang":  bonfy["lang"],
                "days":  app,
                "desc":  bonfy["desc"],
                "rating": rating,
            }
            added += 1

    print(f"\nNovos repos adicionados : {added}")
    print(f"Repos existentes incrementados: {updated}")

    # 4) Aplicar category e recomputar rating para todos os novos -------------
    for r in existing.values():
        r["category"] = categorise(r["name"], r.get("desc", ""))
        # Só recomputa rating para repos sem histórico longo (novos do bonfy)
        if r.get("days", 0) <= 15 and r.get("rating") in (None, ""):
            d = r["days"]
            r["rating"] = "High" if d >= 5 else ("Medium" if d >= 2 else "Low")

    # 5) Re-rankear por days desc ---------------------------------------------
    merged = sorted(existing.values(), key=lambda r: (-r.get("days", 0), r["name"].lower()))
    for i, r in enumerate(merged, 1):
        r["rank"] = i

    # Estatísticas de linguagem
    lang_count = Counter(r["lang"] for r in merged)
    print(f"\nDataset final: {len(merged)} repos")
    for lang, n in sorted(lang_count.items(), key=lambda x: -x[1]):
        print(f"  {lang:<12} {n:>5}")
    go_swift = sum(n for lang, n in lang_count.items()
                   if lang.lower() in ("go", "swift"))
    print(f"  Go + Swift   {go_swift:>5}  (deve ser 0)")

    # 6) Salvar ---------------------------------------------------------------
    json_text = json.dumps(merged, ensure_ascii=False, indent=2)
    REPOS_JSON.parent.mkdir(parents=True, exist_ok=True)
    REPOS_JSON.write_text(json_text, encoding="utf-8")
    print(f"\nSalvo: {REPOS_JSON}  ({REPOS_JSON.stat().st_size // 1024} KB)")

    DOCS_JSON.parent.mkdir(parents=True, exist_ok=True)
    DOCS_JSON.write_text(json_text, encoding="utf-8")
    print(f"Salvo: {DOCS_JSON}  ({DOCS_JSON.stat().st_size // 1024} KB)")

    # 7) Regenerar artefatos web (data.js + index.html) -----------------------
    print("\nRegenerando docs/data.js e docs/index.html ...")
    subprocess.run(
        [sys.executable, "scripts/generate_data.py", "repos.json", "data/repos.json"],
        cwd=ROOT, check=True
    )
    print("\nConcluído.")

if __name__ == "__main__":
    main()

File: scripts/fetch_trending.js
#!/usr/bin/env node
/**
 * fetch_trending.js
 *
 * Busca os últimos N dias do bonfy/github-trending, filtra Python/JavaScript,
 * mescla com o dataset existente (docs/data/repos.json) e regenera
 * docs/data/repos.json, data/repos.json e docs/data.js.
 *
 * Uso: node scripts/fetch_trending.js [--days 15]
 */

import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT      = resolve(__dirname, '..');

const REPOS_JSON = resolve(ROOT, 'data/repos.json');
const DOCS_JSON  = resolve(ROOT, 'docs/data/repos.json');
const DATA_JS    = resolve(ROOT, 'docs/data.js');

const ALLOWED_LANGS = new Set(['python', 'javascript']);
const LANG_CANONICAL = { python: 'Python', javascript: 'Javascript' };

// ── categorias (idêntico ao generate_data.py) ─────────────────────────────
const CATEGORIES = [
    ['Agentic AI', ['agent','crewai','autogen','langgraph','mcp','tool-calling','tool_calling','autonomous','multi-agent','multiagent','agentic']],
    ['LLMs & Models', ['llm','gpt','claude','llama','mistral','gemini','fine-tun','finetun','gguf','ollama','ggml','transformers','huggingface','openai','chatgpt','anthropic','deepseek','qwen','falcon','vllm','lora','rlhf']],
    ['Image & Video', ['diffusion','comfyui','stable-diffusion','stablediffusion','image generation','text-to-image','text-to-video','animate','gan','inpaint','controlnet','dreambooth','video generation','img2img']],
    ['Audio & Voice', ['tts','whisper','speech','voice','audio','music','text-to-speech','stt','asr','transcri','vocal']],
    ['Data & Analytics', ['pandas','etl','dbt','kafka','airflow','polars','spark','databricks','bigquery','redshift','snowflake','pyspark','data analysis','data cleaning','sql','warehouse','analytics','matplotlib','plotly','jupyter']],
    ['Security', ['pentest','osint','ctf','vulnerability','forensic','exploit','hacking','malware','reverse engineer','cybersecurity','infosec','owasp','xss','injection','zero-day']],
    ['Games & Creative', ['game','3d','pixel','canvas','shader','graphics','opengl','unity','godot','pygame','raylib','threejs','creative coding']],
    ['Self-hosted', ['self-host','selfhost','homelab','home assistant','nas','nextcloud','jellyfin','plex','immich','download manager','paperless','torrent','media server']],
    ['Automation & Bots', ['bot','automat','workflow','n8n','selenium','playwright','rpa','scraper','crawler','zapier','webhook','scheduler']],
    ['Learning', ['tutorial','course','roadmap','awesome','interview','book','guide','learn','beginner','30-days','cheatsheet','primer','from scratch','algorithms','data structure','system design','leetcode']],
    ['DevTools', ['cli','docker','kubernetes','k8s','lint','test','debug','monitor','infra','devops','terraform','helm','ansible','formatter','linter','vscode','vim','neovim','shell','zsh','ci/cd','github action','build tool']],
    ['Web & APIs', ['fastapi','django','flask','react','next.js','nextjs','vue','svelte','angular','graphql','nuxt','tailwind','bootstrap','express','nodejs','rest api','frontend','backend','fullstack','typescript','web framework']],
];

function categorise(name, desc) {
    const text = `${name} ${desc || ''}`.toLowerCase();
    for (const [label, kws] of CATEGORIES) {
        if (kws.some(kw => text.includes(kw))) return label;
    }
    return 'Other';
}

// ── URL do arquivo bonfy ──────────────────────────────────────────────────
function bonfy_url(d) {
    const s = d.toISOString().slice(0, 10);   // YYYY-MM-DD
    return d.getFullYear() >= 2026
        ? `https://raw.githubusercontent.com/bonfy/github-trending/master/${s}.md`
        : `https://raw.githubusercontent.com/bonfy/github-trending/master/${d.getFullYear()}/${s}.md`;
}

async function fetchRaw(url) {
    try {
        const res = await fetch(url, { headers: { 'User-Agent': 'fetch_trending/1.0' } });
        if (res.ok) return await res.text();
        if (res.status !== 404) console.error(`  HTTP ${res.status}: ${url}`);
    } catch (e) {
        console.error(`  Erro ao buscar ${url}: ${e.message}`);
    }
    return null;
}

// ── parser do formato bonfy ───────────────────────────────────────────────
// Cada repo: * [owner / repo](url):descrição
const REPO_RE = /^\* \[([^\]]+)\]\(([^)]+)\)(?::(.*))?$/;

function parseBonfyMd(content) {
    const repos = [];
    let lang = null;
    for (const rawLine of content.split('\n')) {
        const line = rawLine.trimEnd();
        if (line.startsWith('#### ')) {
            lang = line.slice(5).trim().toLowerCase();
        } else if (line.startsWith('* [') && lang && ALLOWED_LANGS.has(lang)) {
            const m = REPO_RE.exec(line);
            if (!m) continue;
            const name = m[1].trim().replace(/\s*\/\s*/g, '/');
            const url  = m[2].trim();
            const desc = (m[3] || '').trim();
            repos.push({ name, url, desc, lang: LANG_CANONICAL[lang] });
        }
    }
    return repos;
}

// ── argparse básico ───────────────────────────────────────────────────────
function getArgs() {
    const args = process.argv.slice(2);
    const idx  = args.indexOf('--days');
    return { days: idx >= 0 ? parseInt(args[idx + 1], 10) : 15 };
}

// ── main ──────────────────────────────────────────────────────────────────
async function main() {
    const { days } = getArgs();
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    const dates = Array.from({ length: days }, (_, i) => {
        const d = new Date(today);
        d.setDate(d.getDate() - (days - 1 - i));   // cronológico
        return d;
    });

    // 1) Buscar bonfy ─────────────────────────────────────────────────────
    console.log(`Buscando ${days} dias do bonfy/github-trending...`);
    const bonfyMap = new Map();   // url → {name, url, lang, desc, appearances}
    let fetched = 0;

    for (const d of dates) {
        const url     = bonfy_url(d);
        const content = await fetchRaw(url);
        if (!content) {
            console.log(`  ${d.toISOString().slice(0,10)}  404/erro — ignorado`);
            continue;
        }
        const repos = parseBonfyMd(content);
        fetched++;
        console.log(`  ${d.toISOString().slice(0,10)}  ${String(repos.length).padStart(3)} repos (Python/JS)`);

        for (const r of repos) {
            if (bonfyMap.has(r.url)) {
                bonfyMap.get(r.url).appearances++;
                if (r.desc && !bonfyMap.get(r.url).desc) bonfyMap.get(r.url).desc = r.desc;
            } else {
                bonfyMap.set(r.url, { ...r, appearances: 1 });
            }
        }
    }

    console.log(`\nTotal de datas buscadas: ${fetched}/${days}`);
    console.log(`Repos únicos Python/JS no período: ${bonfyMap.size}`);

    // Safeguard: eliminar Go/Swift caso apareçam
    let goSwiftSkipped = 0;
    for (const [url, r] of bonfyMap) {
        if (!['Python', 'Javascript'].includes(r.lang)) { bonfyMap.delete(url); goSwiftSkipped++; }
    }
    if (goSwiftSkipped) console.log(`Repos Go/Swift bloqueados: ${goSwiftSkipped}`);

    // 2) Carregar dataset existente ───────────────────────────────────────
    let existing = new Map();
    try {
        const raw = readFileSync(DOCS_JSON, 'utf8');
        for (const r of JSON.parse(raw)) existing.set(r.url, r);
    } catch {
        try {
            const raw = readFileSync(REPOS_JSON, 'utf8');
            for (const r of JSON.parse(raw)) existing.set(r.url, r);
        } catch { /* dataset vazio */ }
    }
    console.log(`Repos existentes no dataset: ${existing.size}`);

    // 3) Mesclar ──────────────────────────────────────────────────────────
    let added = 0, updated = 0;
    for (const [url, bonfy] of bonfyMap) {
        const app = bonfy.appearances;
        if (existing.has(url)) {
            const r = existing.get(url);
            r.days = (r.days || 0) + app;
            if (bonfy.desc && !r.desc) r.desc = bonfy.desc;
            updated++;
        } else {
            existing.set(url, {
                name:   bonfy.name,
                url,
                lang:   bonfy.lang,
                days:   app,
                desc:   bonfy.desc,
                rating: app >= 5 ? 'High' : app >= 2 ? 'Medium' : 'Low',
            });
            added++;
        }
    }

    console.log(`\nNovos repos adicionados       : ${added}`);
    console.log(`Repos existentes incrementados : ${updated}`);

    // 4) Atualizar category e rating nos novos ────────────────────────────
    for (const r of existing.values()) {
        r.category = categorise(r.name, r.desc);
        if ((r.days || 0) <= 15 && !r.rating) {
            r.rating = r.days >= 5 ? 'High' : r.days >= 2 ? 'Medium' : 'Low';
        }
    }

    // 5) Re-rankear por days desc ─────────────────────────────────────────
    const merged = [...existing.values()].sort((a, b) =>
        (b.days || 0) - (a.days || 0) || a.name.toLowerCase().localeCompare(b.name.toLowerCase())
    );
    merged.forEach((r, i) => { r.rank = i + 1; });

    // Estatísticas
    const langCount = {};
    for (const r of merged) langCount[r.lang] = (langCount[r.lang] || 0) + 1;
    console.log(`\nDataset final: ${merged.length} repos`);
    for (const [lang, n] of Object.entries(langCount).sort((a, b) => b[1] - a[1]))
        console.log(`  ${lang.padEnd(12)} ${n}`);
    const goSwift = Object.entries(langCount)
        .filter(([l]) => ['go','swift','Go','Swift'].includes(l))
        .reduce((s, [, n]) => s + n, 0);
    console.log(`  Go + Swift   ${goSwift}  (deve ser 0)`);

    // 6) Salvar ───────────────────────────────────────────────────────────
    const jsonText = JSON.stringify(merged, null, 2);

    mkdirSync(dirname(REPOS_JSON), { recursive: true });
    writeFileSync(REPOS_JSON, jsonText, 'utf8');
    console.log(`\nSalvo: ${REPOS_JSON}  (${Math.round(Buffer.byteLength(jsonText) / 1024)} KB)`);

    mkdirSync(dirname(DOCS_JSON), { recursive: true });
    writeFileSync(DOCS_JSON, jsonText, 'utf8');
    console.log(`Salvo: ${DOCS_JSON}  (${Math.round(Buffer.byteLength(jsonText) / 1024)} KB)`);

    const dataJs = `const REPOS = ${JSON.stringify(merged)};\n`;
    writeFileSync(DATA_JS, dataJs, 'utf8');
    console.log(`Salvo: ${DATA_JS}  (${Math.round(Buffer.byteLength(dataJs) / 1024)} KB)`);

    console.log('\nConcluído.');
}

main().catch(err => { console.error(err); process.exit(1); });
